base/src/commonMain/kotlin/utils/BufferedAsListChannel.kt

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2023 Splendo Consulting B.V. The Netherlands
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ */
+
+package com.splendo.kaluga.base.utils
+
+import com.splendo.kaluga.base.singleThreadDispatcher
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ChannelIterator
+import kotlinx.coroutines.channels.ChannelResult
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.selects.SelectClause1
+import kotlinx.coroutines.selects.SelectClause2
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * A Rendezvous Channel that buffers all elements sent to it in a list until the next receive
+ * @param T the type of element to batch. Must be non-nullable
+ */
+interface BufferedAsListChannel<T : Any> : SendChannel<T>, ReceiveChannel<List<T>>
+
+/**
+ * Creates a [BufferedAsListChannel] that batches its elements with a given [CoroutineContext]
+ * @param T the type of element to batch. Must be non-nullable
+ * @param coroutineContext the [CoroutineContext] to use for batching
+ * @return the [BufferedAsListChannel] created
+ */
+fun <T : Any> BufferedAsListChannel(coroutineContext: CoroutineContext): BufferedAsListChannel<T> = BatchingChannelInt(coroutineContext)
+
+internal class BatchingChannelInt<T : Any>(coroutineContext: CoroutineContext) : BufferedAsListChannel<T> {
+    private val sendChannel = Channel<T>(Channel.UNLIMITED)
+    private val receiveChannel = Channel<List<T>>()
```

Comment:
The tests below for this are quite limited. At the same time, you don't want to waste time spending methods that are just delegated.

I wonder if this could be made more explicit by delegating the implementation of `SendChannel` / `ReceiveChannel ` `by` using interface delegation, and only overriding methods that have different behaviour (e.g. cancel). Then you won't have to look at methods like 
```kotlin
override val onReceiveCatching: SelectClause1<ChannelResult<List<T>>>
        get() = receiveChannel.onReceiveCatching`
```
and obviously not need to test them either.

Not 100% sure this concept will work, but worth exploring.

Reply:
Seems to work 👍